// @shopify/cli@3.80.6/dist/local-US2LLU7Y.js
import {
  alwaysLogAnalytics,
  alwaysLogMetrics,
  analyticsDisabled,
  ciPlatform,
  cloudEnvironment,
  codespacePortForwardingDomain,
  codespaceURL,
  firstPartyDev,
  getThemeKitAccessDomain,
  gitpodURL,
  hasGit,
  homeDirectory,
  isCloudEnvironment,
  isDevelopment,
  isShopify,
  isTerminalInteractive,
  isUnitTest,
  isVerbose,
  macAddress,
  opentelemetryDomain,
  themeToken,
  useEmbeddedThemeCLI,
  useThemebundling
} from "./chunk-7WRUWTUK.js";
import "./chunk-HMDWNGIV.js";
import "./chunk-B5EXYCV3.js";
import "./chunk-75LV6AQS.js";
import "./chunk-ZUCWDIGE.js";
import "./chunk-5AFYDRV4.js";
import "./chunk-3I3GQNEW.js";
import "./chunk-G5R6YD27.js";
import "./chunk-G2ZZKGSV.js";
import "./chunk-OWRPRJTD.js";
import "./chunk-XE5EOEBL.js";
import "./chunk-TJM3YBLQ.js";
import "./chunk-7IK72W75.js";
import "./chunk-CDBXAE2F.js";
import "./chunk-PKR7KJ6P.js";
export {
  alwaysLogAnalytics,
  alwaysLogMetrics,
  analyticsDisabled,
  ciPlatform,
  cloudEnvironment,
  codespacePortForwardingDomain,
  codespaceURL,
  firstPartyDev,
  getThemeKitAccessDomain,
  gitpodURL,
  hasGit,
  homeDirectory,
  isCloudEnvironment,
  isDevelopment,
  isShopify,
  isTerminalInteractive,
  isUnitTest,
  isVerbose,
  macAddress,
  opentelemetryDomain,
  themeToken,
  useEmbeddedThemeCLI,
  useThemebundling
};
