// @shopify/theme-language-server-node@1.8.3/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.startServer = void 0;
const theme_language_server_common_1 = require("@shopify/theme-language-server-common");
const theme_check_docs_updater_1 = require("@shopify/theme-check-docs-updater");
const node_process_1 = require("node:process");
const node_1 = require("vscode-languageserver/node");
const dependencies_1 = require("./dependencies");
function startServer() {
    const connection = (0, node_1.createConnection)(node_process_1.stdin, node_process_1.stdout);
    const log = (message) => console.error(message);
    const themeLiquidDocsManager = new theme_check_docs_updater_1.ThemeLiquidDocsManager(log);
    (0, theme_language_server_common_1.startServer)(connection, {
        // Using console.error to not interfere with messages sent on STDIN/OUT
        log,
        getDefaultTranslationsFactory: dependencies_1.getDefaultTranslationsFactory,
        getDefaultLocaleFactory: dependencies_1.getDefaultLocaleFactory,
        getThemeSettingsSchemaForRootURI: dependencies_1.getThemeSettingsSchemaForRootURI,
        findRootURI: dependencies_1.findRootURI,
        fileExists: dependencies_1.fileExists,
        fileSize: dependencies_1.fileSize,
        filesForURI: dependencies_1.filesForURI,
        loadConfig: dependencies_1.loadConfig,
        themeDocset: themeLiquidDocsManager,
        jsonValidationSet: themeLiquidDocsManager,
    });
}
exports.startServer = startServer;
