// @shopify/theme-language-server-common@1.8.3/dist/completions/providers/HtmlAttributeCompletionProvider.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.HtmlAttributeCompletionProvider = void 0;
const vscode_languageserver_1 = require("vscode-languageserver");
const docset_1 = require("../../docset");
const utils_1 = require("../../utils");
const params_1 = require("../params");
const common_1 = require("./common");
class HtmlAttributeCompletionProvider {
    constructor() { }
    async completions(params) {
        if (!params.completionContext)
            return [];
        const { node, ancestors } = params.completionContext;
        const parentNode = (0, utils_1.findLast)(ancestors, utils_1.isAttrEmpty);
        const grandParentNode = (0, utils_1.findLast)(ancestors, utils_1.isNamedHtmlElementNode);
        if (!node || !parentNode || !grandParentNode) {
            return [];
        }
        if (!(0, utils_1.isTextNode)(node) || !(0, utils_1.isAttrEmpty)(parentNode) || !(0, utils_1.isNamedHtmlElementNode)(grandParentNode)) {
            return [];
        }
        const grandParentNodeName = (0, utils_1.getCompoundName)(grandParentNode);
        const name = node.value;
        const partial = name.replace(params_1.CURSOR, '');
        const options = getOptions(partial, grandParentNodeName);
        return options.sort(common_1.sortByName).map(toCompletionItem);
    }
}
exports.HtmlAttributeCompletionProvider = HtmlAttributeCompletionProvider;
function getOptions(partial, parentNodeName) {
    var _a;
    const tag = docset_1.HtmlData.tags.find((tag) => tag.name === parentNodeName);
    const parentAttributes = (_a = tag === null || tag === void 0 ? void 0 : tag.attributes) !== null && _a !== void 0 ? _a : [];
    return [...parentAttributes, ...docset_1.HtmlData.globalAttributes].filter((x) => x.name.startsWith(partial));
}
function toCompletionItem(tag) {
    return {
        label: tag.name,
        kind: vscode_languageserver_1.CompletionItemKind.Value,
        documentation: {
            kind: 'markdown',
            value: (0, docset_1.renderHtmlEntry)(tag),
        },
    };
}
