// @shopify/theme-check-common@2.9.2/dist/to-source-code.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.toSourceCode = exports.toJSONAST = exports.toLiquidHTMLAST = void 0;
const liquid_html_parser_1 = require("@shopify/liquid-html-parser");
const json_to_ast_1 = __importDefault(require("json-to-ast"));
const types_1 = require("./types");
const error_1 = require("./utils/error");
function toLiquidHTMLAST(source) {
    try {
        return (0, liquid_html_parser_1.toLiquidHtmlAST)(source);
    }
    catch (error) {
        return (0, error_1.asError)(error);
    }
}
exports.toLiquidHTMLAST = toLiquidHTMLAST;
function toJSONAST(source) {
    try {
        return (0, json_to_ast_1.default)(source);
    }
    catch (error) {
        return (0, error_1.asError)(error);
    }
}
exports.toJSONAST = toJSONAST;
function toSourceCode(absolutePath, source, version) {
    const isLiquid = absolutePath.endsWith('.liquid');
    if (isLiquid) {
        return {
            absolutePath: normalize(absolutePath),
            source,
            type: types_1.SourceCodeType.LiquidHtml,
            ast: toLiquidHTMLAST(source),
            version,
        };
    }
    else {
        return {
            absolutePath: normalize(absolutePath),
            source,
            type: types_1.SourceCodeType.JSON,
            ast: toJSONAST(source),
            version,
        };
    }
}
exports.toSourceCode = toSourceCode;
function normalize(path) {
    return path.replace(/\\/g, '/');
}
