// @shopify/theme-language-server-common@1.8.3/dist/documents/DocumentManager.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DocumentManager = void 0;
const theme_check_common_1 = require("@shopify/theme-check-common");
const vscode_languageserver_textdocument_1 = require("vscode-languageserver-textdocument");
const utils_1 = require("../utils");
class DocumentManager {
    constructor() {
        this.sourceCodes = new Map();
    }
    open(uri, source, version) {
        return this.set(uri, source, version);
    }
    change(uri, source, version) {
        return this.set(uri, source, version);
    }
    close(uri) {
        return this.sourceCodes.delete(uri);
    }
    theme(root) {
        return [...this.sourceCodes.entries()]
            .filter(([uri]) => uri.startsWith(root))
            .map(([, sourceCode]) => sourceCode);
    }
    get openDocuments() {
        return [...this.sourceCodes.values()];
    }
    get(uri) {
        return this.sourceCodes.get(uri);
    }
    set(uri, source, version) {
        var _a;
        const absolutePath = (0, utils_1.toAbsolutePath)(uri);
        // We only support json and liquid files.
        if (!/\.(json|liquid)$/.test(absolutePath) || /\.(s?css|js).liquid$/.test(absolutePath)) {
            return;
        }
        const sourceCode = (0, theme_check_common_1.toSourceCode)(absolutePath, source, version);
        this.sourceCodes.set(uri, {
            ...sourceCode,
            textDocument: vscode_languageserver_textdocument_1.TextDocument.create(uri, sourceCode.type, (_a = sourceCode.version) !== null && _a !== void 0 ? _a : 0, sourceCode.source),
            uri,
        });
    }
}
exports.DocumentManager = DocumentManager;
